Move matrix-sdk to ruma monorepo

DevinR528 opened the pull request.

vbrandl committed:
Implement display name resolving

vbrandl committed:
Rename accessor for display name

vbrandl committed:
Fix failing test

vbrandl committed:
Only add name duplicates to the display name map

vbrandl committed:
Add first test for display names

dkasak committed:
Rename display_names -> disambiguated_display_names.

dkasak committed:
Fix Markdown in doc comment.

dkasak committed:
matrix-sdk: Vary sync token with each EventBuilder::build_sync_response call.

This allows us to hold onto an EventBuilder object and use it to build
multiple sync responses. Previously this would have resulted in each
of the responses having the same next_batch sync token. This would make
clients ignore the latter responses if they have already received any of
the previous ones.

dkasak committed:
Properly test for display name disambiguation.

dkasak committed:
Fix display name disambiguation so it passes the test.

dkasak committed:
Use "disambiguated" instead of "resolved" display name in the doc comment.

To match how the C2S spec calls it.

dkasak committed:
Add short explanation to Room::member_display_name.

dkasak committed:
Make Room::member_display_name return MXID as fallback.

If there is no display name set. This means the method can now always
return something so there is no need to wrap in an `Option`.

dkasak committed:
Run cargo fmt.

dkasak committed:
Add explicit type annotation.

dkasak committed:
Fix comment styling.

dkasak committed:
fixup: explicit type annotations

dkasak committed:
Retrieve user id from RoomMember instead of reconstructing.

dkasak committed:
Refactor member_display_name.

Make it more readable, add comments.

dkasak committed:
Add doc comment to RoomName::calculate_name.

dkasak committed:
Small refactor so we don't duplicate user_id creation.

dkasak committed:
Style fixes.

dkasak committed:
Add RoomMember::name.

Returns the most ergonomic name for the member (either the display name
(if set) or the MXID).

dkasak committed:
Implement RoomMember::unique_name.

This gives us a name that is as ergonomic as possible while guaranteeing
it is unique.

dkasak committed:
Nix RoomMember::update_member and tracking membership.

After discussing with poljar, we concluded we don't actually need to
tracking membership state, since we won't be tracking users that
left (banned, kicked, disinvited).

The only thing we need to keep track of is the difference between joined
and invited users which will be dealt with in a separate commit.

dkasak committed:
Split joined/invited users and handle removing users.

dkasak committed:
Must not take our user into account when calculating room name.

dkasak committed:
Hoist prev_content to top-level in both timeline and state events.

Also refactor and document why this hoisting is needed.

This change makes the user_presence test fail because the hoisting
exposes an error encoded into the test's expected result.

Previously, the test expected 2 members in the room at the end. This is
incorrect since one of the members in the test data leaves the room.
However, since the prev_content of state events was previously not
hoisted to the top level, the `membership_change` method would not
notice it and thus not realize the member had left the room. The test
was corrected to expect only a single member in the room.

Another test change was made due to a limitation of EventBuilder: due to
the fact that it makes the test data go through a de/ser cycle, it
cannot easily hoist prev_content to the top level. Because of this, the
tests were change to put prev_content into the top level from the
outset.

dkasak committed:
Fix comment style.

dkasak committed:
Correctly handle disambiguation for exiting members, refactor and test.

dkasak committed:
Fix compilation error and remaining test.

Ref. for compilation error:
https://github.com/rust-lang/rust/issues/64552

dkasak committed:
Merge branch 'master' into feature/display-name

dkasak committed:
Run cargo fmt and apply clippy lints.

DevinR528 committed:
Move matrix-sdk to ruma monorepo

DevinR528 committed:
matrix_sdk: Add get_public_rooms* methods to Client

This also adds a RoomSearchBuilder for making get_public_rooms_filtered
requests and a test for each method.

poljar committed:
matrix-sdk: Remove the last test_data folder and fix the remaining tests.

poljar committed:
Merge branch 'feature/display-name'

poljar committed:
matrix-sdk: Disable the tarpaulin skip lines since it fails to run with them.

DevinR528 committed:
matrix_sdk: Rename public room builder and client methods

Remove 'get' from get_public_rooms* methods.
Rename RoomSearchBuilder -> RoomListFilterBuilder.
Use u32 over UInt in builders and Into<String> for String.
Fix docs of public room methods and builders.

poljar commented on matrix_sdk_base/src/client.rs:
```diff
@@ -1405,42 +1420,58 @@ impl BaseClient {
+                    event_emitter.on_room_tombstone(room, &e).await
+                }
+                AnyStateEventStub::RoomJoinRules(_e) => {
+                    // TODO is this needed ??
```
I would say it is, consider a client like Riot. Riot changes the decoration based on the join rule, if the room is invite only it gets decorated with a lock.

poljar commented on matrix_sdk_base/src/models/room.rs:
```diff
@@ -21,27 +21,34 @@ use super::message::MessageQueue;
 use serde::{Deserialize, Serialize};
+
+#[cfg(feature = "messages")]
+#[allow(dead_code)]
```
Why is this `allow` needed?

poljar commented on matrix_sdk_base/src/models/room.rs:
```diff
@@ -491,21 +505,28 @@ impl Room {
+            AnyRoomEventStub::Message(event) => match &event {
+                #[cfg(feature = "messages")]
+                AnyMessageEventStub::RoomMessage(event) => self.handle_message(&event),
+                // TODO if a redaction event deletes one of our saved messages delete it?
```
Probably better to modify it to look like a redacted event, that is remove the content otherwise you'll end  up with a gap instead of showing a redacted event when you restore the messages.

DevinR528 replied:
We would have to turn the MessageWrapper thing that holds MessageEventStub<MessageEventContent> to an enum to hold the messages and `RedactionEventStub`, they are separate types unfortunatly, they are both AnyMessageEvntStub that would work maybe.

poljar reviewed (approved):
This looks mostly good, I'm still not sure what to do with the normal event vs stub event duplication, perhaps just add two decrypt methods but we can figure that out later on.

DevinR528 committed:
Use ruma/ruma master, address review issues

DevinR528 committed:
Rebase upstream/master into ruma-mono branch

DevinR528 committed:
matrix-sdk-base: Use new accessor methods for models/message.rs

ruma now has field access methods for all of the Any*Event enums use
them for MessageWrapper's AnyMessageEventStub contents.

poljar committed:
crypto: Simplify the group session pair creation.

DevinR528 committed:
Move test data to test crate, fix docs

poljar committed:
Merge branch 'room-search'

poljar committed:
matrix-sdk: Remove an unused import.

poljar committed:
base: Put the decrypted event replacing in the correct place.

dkasak committed:
Add a bunch of FIXMEs to have receive_* methods do the emitting.

dkasak committed:
Fix and test hoisting of `prev_content` for timeline events.

The previous test only tested using the `EventEmitter`, which missed the
fact that the client was receiving unhoisted events. The test now also
tests the client state to detect this.

poljar committed:
base: Don't double borrow the response in one iter rooms method.

DevinR528 committed:
matrix-sdk-base: clean up recv/iter joined post rebase

The types for account data in a sync response have changed, no longer
Option. Re word comment in hoist prev_content test.
